shell/platform/embedder/embedder.h

```diff
@@ -1118,9 +1121,82 @@ typedef struct {
   const char* tooltip;
 } FlutterSemanticsNode;
 
+/// A node that represents some semantic data.
+///
+/// The semantics tree is maintained during the semantics phase of the pipeline
+/// (i.e., during PipelineOwner.flushSemantics), which happens after
+/// compositing. Updates are then pushed to embedders via the registered
+/// `FlutterUpdateSemanticsCallback2`.
+typedef struct {
```

Comment:
What is the reason for defining this new struct to use instead of the existing one? Is this only to keep the old and new ABIs separated?

Reply:
This is necessary as `FlutterSemanticsUpdate` has an array of `FlutterSemanticsNode` and an array of `FlutterSemanticsCustomAction`. Adding new members to these structs would break the ABI compatibility of array indexing. In other words, `FlutterSemanticsNode` and `FlutterSemanticsCustomAction` must never change.

However, `FlutterSemanticsUpdate2` has array of pointers to `FlutterSemanticsNode2` and an array of pointers to `FlutterSemanticsCustomAction2`. It is safe to add new members to `FlutterSemanticsNode2` and `FlutterSemanticsCustomAction2`. For more information, please see the Background section of https://github.com/flutter/flutter/issues/121176 or see [go/flutter-semantics-abi](http://go/flutter-semantics-abi).